Jdbc merge support

chenjian2664 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Close #20532

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# PostgresSQL Jdbc
* Adding non-transactional support for {doc}`MERGE statements </sql/merge>`.

```

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+        this.deleteSink = createDeleteSink(session, jdbcMergeTableHandle.getDeleteTableHandle(), jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    private List<Integer> getReservedChannelsForUpdate(JdbcOutputTableHandle outputTableHandle)
```
you are calling a method in constructor, the methods is using fields of the object. So it requires specific order of statements in constructor to make sure this method works correctly. Consider using static method here.

kokosing replied:
`getReservedChannelsForUpdate` -> `getNonMergeIdChannelsForUpdate`?

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+
+    private List<Integer> getReservedChannelsForUpdate(JdbcOutputTableHandle outputTableHandle)
+    {
+        Set<String> excludedColumnNames = ImmutableSet.copyOf(mergeRowIdFieldNames);
```
`excludedColumnNames` -> `mergeRowIdFieldNames`?

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+        this.reservedChannelsForUpdate = getReservedChannelsForUpdate(jdbcOutputTableHandle);
+        verify(!reservedChannelsForUpdate.isEmpty(), "Update primary key itself is not supported");
+
+        this.pageSinkId = pageSinkId;
```
requireNonNull

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+        return reservedChannelsBuilder.build();
+    }
+
+    protected ConnectorPageSink createInsertSink(
```
inline?

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+        return new JdbcPageSink(session, jdbcOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    protected ConnectorPageSink createUpdateSink(
```
same comment about static methods called in constructor

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+        return new UpdateSink(session, updateOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    protected ConnectorPageSink createDeleteSink(
```
same comment about static methods called in constructor

chenjian2664 replied:
If make this method to static how to override it. Phoenix needs to create a different `updateSink`

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+        updateColumnNames.addAll(mergeRowIdFieldNames);
+        updateColumnTypes.addAll(mergeRowIdFieldTypes);
+
+        JdbcOutputTableHandle updateOutputTableHandle = new JdbcOutputTableHandle(
```
Why can't we use `jdbcMergeTableHandle.getOutputTableHandle()` that we got in constructor?

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeTableHandle.java:
```diff
@@ -0,0 +1,78 @@
+{
+    private final JdbcTableHandle tableHandle;
+    private final JdbcOutputTableHandle outputTableHandle;
+    private final Optional<JdbcOutputTableHandle> deleteTableHandle;
```
This field is only used when you add support for merge to postgres, so I believe this is when this field should be added.

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+        @Override
+        protected String getSinkSql(JdbcClient jdbcClient, JdbcOutputTableHandle outputTableHandle, List<WriteFunction> columnWriters)
+        {
+            return "DELETE FROM %s.%s WHERE %s".formatted(jdbcClient.quoted(schemaName), jdbcClient.quoted(tableName), mergeRowIdConjuncts);
```
I believe we should somehow reuse `io.trino.plugin.jdbc.DefaultQueryBuilder#prepareDeleteQuery`

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+        }
+
+        @Override
+        protected String getSinkSql(JdbcClient jdbcClient, JdbcOutputTableHandle outputTableHandle, List<WriteFunction> columnWriters)
```
I think we should somehow reuse `io.trino.plugin.jdbc.DefaultQueryBuilder#prepareUpdateQuery`

kokosing commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteMergeTableHandle.java:
```diff
@@ -0,0 +1,33 @@
+import io.trino.plugin.jdbc.JdbcMergeTableHandle;
+import io.trino.plugin.jdbc.JdbcTableHandle;
+
+public class IgniteMergeTableHandle
```
Why this class is added? It looks like plain `JdbcMergeTableHandle`

chenjian2664 replied:
Needs to hold `IgniteOutputTableHandle`

kokosing commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteMetadata.java:
```diff
@@ -123,6 +130,25 @@ public Optional<ConnectorOutputMetadata> finishInsert(
+    @Override
+    public Optional<ConnectorTableHandle> applyUpdate(ConnectorSession session, ConnectorTableHandle handle, Map<ColumnHandle, Constant> assignments)
+    {
+        // Ignite support row level update, so we should reject this path, earlier than in JDBC client
```
Why? If possible we should use `applyUpdate` as this could way more faster. What am I missing?

kokosing commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteMetadata.java:
```diff
@@ -123,6 +130,25 @@ public Optional<ConnectorOutputMetadata> finishInsert(
+    }
+
+    @Override
+    public void finishMerge(ConnectorSession session, ConnectorMergeTableHandle mergeTableHandle, List<ConnectorTableHandle> sourceTableHandles, Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics)
```
Why do we have customize here?

chenjian2664 replied:
Ignite not support transaction write, no finish operation required.

kokosing commented on plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -1156,6 +1160,41 @@ protected boolean isSupportedJoinCondition(ConnectorSession session, JdbcJoinCon
+    }
+
+    @Override
+    public JdbcOutputTableHandle beginDeleteTableForMerge(ConnectorSession session, JdbcTableHandle tableHandle)
```
As I understand this is the only JDBC connector (that supports this method) so why not have `BaseJdbcClient` to be written based on Postgres?

chenjian2664 replied:
Postgres maybe the first connector but not the last. The connectors that support transaction insert could support MERGE that would require to implement this.

chenjian2664 replied:
Postgres has it's own special key `ctid`, we should avoid refer it in the `BaseJdbcClient`

kokosing reviewed (commented):
Another round of review. Please have a lot of patience, I still try to get deeper and still I am not there. There could be a case where some my comments could be reverted in future. It is a creative process, so I hope in the end we will get into a good spot with this PR, but it may require trials and errors.

I mostly reviewed the first commit. I asked for some preparatory commits, potentially we could do them as separate PR and merge it faster. Then we could rebase this PR.

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcClient.java:
```diff
@@ -258,6 +260,15 @@ default Optional<TableScanRedirectApplicationResult> getTableScanRedirection(Con
 
     OptionalInt getMaxWriteParallelism(ConnectorSession session);
 
+    JdbcTableHandle updatedScanColumnsForMerge(ConnectorSession session, ConnectorTableHandle table, Optional<List<JdbcColumnHandle>> originalColumns, JdbcColumnHandle mergeRowIdColumnHandle);
```
Thanks.
I moved it to the `JdbcPageSourceProvider` since I found we do not need it in the page sink. 
I updated the logic in `JdbcPageSourceProvider`, basicly we want to replace the merge id with its fields, then adapt to the upstream ` columns`, now it seems cleaner.

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcClient.java:
```diff
@@ -258,6 +260,15 @@ default Optional<TableScanRedirectApplicationResult> getTableScanRedirection(Con
+
+    String buildMergeRowIdConjuncts(ConnectorSession session, List<String> mergeRowIdFieldNames, List<Type> mergeRowIdFieldTypes);
+
+    default List<JdbcColumnHandle> getPrimaryKeys(ConnectorSession session, JdbcTableHandle tableHandle)
```
> Should we rename this method to make it clear it is part of the merge?

Do you have any suggestion? `getPrimaryKeysForMerge`?

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcClient.java:
```diff
@@ -258,6 +260,15 @@ default Optional<TableScanRedirectApplicationResult> getTableScanRedirection(Con
 
+    JdbcTableHandle updatedScanColumnsForMerge(ConnectorSession session, ConnectorTableHandle table, Optional<List<JdbcColumnHandle>> originalColumns, JdbcColumnHandle mergeRowIdColumnHandle);
+
+    String buildMergeRowIdConjuncts(ConnectorSession session, List<String> mergeRowIdFieldNames, List<Type> mergeRowIdFieldTypes);
```
The method is override in third commit, Postgres needs it to cast the column to `ctid` type.

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -1774,6 +1778,88 @@ private static RemoteTableName getRemoteTable(ResultSet resultSet)
+
+    protected List<JdbcColumnHandle> getUpdatedScanColumnHandles(ConnectorSession session, JdbcTableHandle tableHandle, List<JdbcColumnHandle> scanColumnHandles, JdbcColumnHandle mergeRowIdColumnHandle)
+    {
+        RowType columnType = (RowType) mergeRowIdColumnHandle.getColumnType();
```
Do you mean `checkArgument(mergeRowIdColumnHandle.getColumnType() instance of RowType)` ?

chenjian2664 commented on plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java:
```diff
@@ -13,216 +13,75 @@
-                }
-                default -> throw new IllegalStateException("Unexpected value: " + operation);
-            }
+        if (!hasRowKey) {
```
> when it happens that row key is absent?

There will be rowkeys defined by the user not added automatically by Trino.

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,309 @@
+{
+    private final String catalogName;
+    private final String schemaName;
+    private final String tableName;
```
filed https://github.com/trinodb/trino/pull/23090

chenjian2664 committed:
Support MERGE in base jdbc module

Support MERGE in base jdbc module and refactor Phoenix implementation based on it

chenjian2664 committed:
Support merge in Postgresql connector

Co-Authored-By: Grzegorz Kokosiński <7569403+kokosing@users.noreply.github.com>

chenjian2664 commented:
Add support for postgresql.
Separate Fte with merge on https://github.com/trinodb/trino/issues/23345

chenjian2664 commented:
It's maybe a bug https://github.com/trinodb/trino/actions/runs/11322914792/job/31484601352?pr=23034#step:5:8834

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,322 @@
+        ImmutableList.Builder<JdbcAssignmentItem> assignmentItemBuilder = ImmutableList.builder();
+        ImmutableList.Builder<String> columnNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> columnTypesBuilder = ImmutableList.builder();
+        QueryParameter dummy = new QueryParameter(BIGINT, Optional.empty());
```
This requires a comment like above with tuple domain. Let's write that value won't be used as the real value will be passed to `storeMergedRows`

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,322 @@
+
+        // The TupleDomain for building the conjuncts of the primary keys
+        ImmutableMap.Builder<ColumnHandle, Domain> primaryKeysDomainBuilder = ImmutableMap.builder();
+        // This value is to build the TupleDomain, but it won't affect the query field in result of the `QueryBuilder#prepareDeleteQuery` and `QueryBuilder#prepareUpdateQuery`
```
This comment needs a rewrite. Let's write that value won't be used as the real value will be passed to `storeMergedRows`

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,322 @@
+        // The TupleDomain for building the conjuncts of the primary keys
+        ImmutableMap.Builder<ColumnHandle, Domain> primaryKeysDomainBuilder = ImmutableMap.builder();
+        // This value is to build the TupleDomain, but it won't affect the query field in result of the `QueryBuilder#prepareDeleteQuery` and `QueryBuilder#prepareUpdateQuery`
+        Domain dummy = Domain.singleValue(BIGINT, 0L);
```
What if the column is not compatible with BIGINT? What if VARCHAR column being used as primary index?

chenjian2664 replied:
The value won't be used in the `DefaultQueryBuilder#prepareUpdateQuery`. in `DefaultQueryBuilder#toConjuncts`, if the domain is a single value, `toPredicate` would build rely on `column` and `writeFunction`.

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,322 @@
+                            domain,
+                            Optional.empty(),
+                            assignmentItems)
+                    .query();
```
I don't think you should use simple query like that. I think you should use parameters too do a column mapping. Now you have tight coupling between this class and query builder. And coupling is not on contract level, but on implementation details.

In `io.trino.plugin.jdbc.JdbcClient#buildInsertSql` is a bit better as order matches the order of passed arguments, but assuming that assignments are handled before domain is a bit risky long term.

chenjian2664 replied:
I am not sure I understand. Do you mean it is better to not return query and instead should return a PreparedQuery here, then use it somehow maybe like `DefaultQueryBuilder#prepareStatement` to handle the mapping?

chenjian2664 replied:
We can not get the value for parameters on constructor when build the `updateSink` or `deleteSink`, and if we want to use parameters then probably not able to reuse `JdbcPageSink`, maybe create a class called JdbcUpdatePageSink?

chenjian2664 replied:
Or do you mean I need to build the query directly, not use the `DefaultQueryBuilder`?

kokosing replied:
> Do you mean it is better to not return query and instead should return a PreparedQuery here, then use it somehow maybe like DefaultQueryBuilder#prepareStatement to handle the mapping?

Yes. `QueryBuilder` is designed to handle single list of parameters. It works well for `SELECT` queries. However here we are using batch of queries. I wonder what it would take to change `QueryBuilder` to be able to support batch too. Then maybe `QueryParameter` should not have a value, but some kind of a "placeholder" which could be then filled for `SELECT` kind of queries or batches.

I guess we are going out of the scope of this PR. Maybe we could do it as follow up.

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,322 @@
+            for (int channel = 0; channel < updateAssignmentChannels.length; channel++) {
+                updateBlocks[channel] = dataPage.getBlock(updateAssignmentChannels[channel]).getPositions(updatePositions, 0, updatePositionCount);
+            }
+            for (int field = 0; field < rowIdFields.size(); field++) {
```
This is seems to the for loop above. Maybe a method extract?

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeTableHandle.java:
```diff
@@ -0,0 +1,61 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class JdbcMergeTableHandle
```
a record?

chenjian2664 replied:
The PhoenixMergeTableHandle dep on this, seems can not convert to record.

kokosing reviewed (commented):
I am in the half of the first commit.

After the refactors you have done, it looks way better now!

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSourceProvider.java:
```diff
@@ -0,0 +1,135 @@
+                tableHandle.getRequiredNamedRelation().getSchemaTableName(),
+                tableHandle.getRequiredNamedRelation().getRemoteTableName(),
+                Optional.empty());
+        List<JdbcColumnHandle> columns = jdbcClient.getColumns(session, plainTable);
```
yet another example of using plain table, that could benefit from the refactor in https://github.com/trinodb/trino/pull/23034/commits/fed8ee4ad8556f4ad641a592a2fb55691227d75f#r1799571242

kokosing replied:
rename columns to allTableColumns

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSourceProvider.java:
```diff
@@ -0,0 +1,135 @@
+            return new RecordPageSource(recordSetProvider.getRecordSet(transaction, session, split, tableHandle, columns));
+        }
+
+        List<JdbcColumnHandle> primaryKeys = jdbcClient.getPrimaryKeys(session, tableHandle.getRequiredNamedRelation().getRemoteTableName());
```
wrap all the below as a method `createMergePageSource`

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSourceProvider.java:
```diff
@@ -0,0 +1,135 @@
+        ImmutableList.Builder<JdbcColumnHandle> scanColumnsBuilder = ImmutableList.builder();
+        scanColumnsBuilder.addAll(columns);
+        // Add merge row id fields
+        for (JdbcColumnHandle column : primaryKeys) {
```
s/column/primaryKey

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSourceProvider.java:
```diff
@@ -0,0 +1,135 @@
+                tableHandle.getRequiredNamedRelation().getSchemaTableName(),
+                tableHandle.getRequiredNamedRelation().getRemoteTableName(),
+                Optional.empty());
+        List<JdbcColumnHandle> columns = jdbcClient.getColumns(session, plainTable);
```
Why to get columns again, merge id column should already have all the information like column and type. What else do you need here?

chenjian2664 replied:
It's easier to get the `JdbcColumnHandle` and by pass it for the Set.contains, List.indexOf, otherwise using names, that not clean.

kokosing reviewed (commented):
Before looking at 3rd commit

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSource.java:
```diff
@@ -0,0 +1,143 @@
+import static io.trino.spi.block.RowBlock.fromFieldBlocks;
+import static java.util.Objects.requireNonNull;
+
+public class JdbcPageSource
```
How does it differ from io.trino.plugin.base.MappedPageSource?

> When the merge rowId has more than one field, the merge rowId column is mapped to multi source, but the MappedPageSource only support one-one mapping if I understand correctly.

Renamed to `MergeJdbcPageSource`

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSource.java:
```diff
@@ -0,0 +1,143 @@
+    {
+        Block getBlock(Page sourcePage);
+
+        static ColumnAdaptation sourceColumn(int index)
```
`SourceColumn` is a private inner record, maybe it's better to keep it?

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/MergeJdbcPageSource.java:
```diff
@@ -0,0 +1,143 @@
+    {
+        Block getBlock(Page sourcePage);
+
+        static ColumnAdaptation sourceColumn(int index)
```
Please inline these static methods. They don't help much.

kokosing commented on plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -1629,6 +1639,29 @@ private static ObjectWriteFunction vectorWriteFunction()
 
+    private static ColumnMapping ctidColumnMapping()
+    {
+        return ColumnMapping.sliceMapping(
```
Let's define a constant for this.

kokosing commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/BasePostgresFailureRecoveryTest.java:
```diff
@@ -81,4 +81,20 @@ protected void testUpdate()
+    protected void testMerge()
+    {
+        // TODO: support merge with fte
+        abort("skipped");
```
please assert the failure of the test. I would like to see that it truly fails because of fte

kokosing commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:
```diff
@@ -139,6 +142,13 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
+    @Override
+    public void testUpdateRowConcurrently()
+    {
+        abort("Postgresql doesn't support concurrent update of different columns in a row");
```
How does it fail then?

chenjian2664 replied:
There will be assertion error, the expected row not matches.
e.g, there are two columns: c1, c2 in the row, starts: `[0,0]`, update them +1 separately, expected : `[1, 1]`, but current implement would override the whole row in a update, it's possible result in `[0,1]` or `[1, 0]` if concurrent update.

kokosing commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java:
```diff
@@ -1882,7 +1882,7 @@ public void testConstantUpdateWithVarcharGreaterAndLowerPredicate()
         skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE) && hasBehavior(SUPPORTS_UPDATE));
         try (TestTable table = new TestTable(getQueryRunner()::execute, "test_update_varchar", "(col1 INT, col2 varchar(1))", ImmutableList.of("1, 'a'", "2, 'A'"))) {
-            if (!hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY)) {
+            if (!hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY) && !hasBehavior(SUPPORTS_MERGE)) {
```
did you mean `SUPPORTS_ROW_LEVEL_UPDATE`?

Should you remove `!hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY)` condition?

chenjian2664 replied:
Renamed.

> Should you remove !hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY) condition?

Can't remove here, the logic here is if either the connector has behavior `SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY` or `SUPPORTS_ROW_LEVEL_UPDATE` the  connector can pushdown the update with constant varchar.

kokosing commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java:
```diff
@@ -1867,7 +1867,7 @@ public void testConstantUpdateWithVarcharInequalityPredicates()
         skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE) && hasBehavior(SUPPORTS_UPDATE));
         try (TestTable table = new TestTable(getQueryRunner()::execute, "test_update_varchar", "(col1 INT, col2 varchar(1))", ImmutableList.of("1, 'a'", "2, 'A'"))) {
-            if (!hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY)) {
+            if (!hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY) && !hasBehavior(SUPPORTS_MERGE)) {
```
did you mean `SUPPORTS_ROW_LEVEL_UPDATE`?

Should you remove `!hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_INEQUALITY)` condition?

kokosing reviewed (commented):
Looks good. Do you think the getColumns refactor should go before or after this PR?

kokosing commented:
@hashhar do you want to take a look? I think it is getting good to go.

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java:
```diff
@@ -0,0 +1,322 @@
+            RemoteQueryModifier remoteQueryModifier,
+            QueryBuilder queryBuilder)
+    {
+        List<JdbcColumnHandle> columns = jdbcClient.getColumns(session, new JdbcTableHandle(relation.getSchemaTableName(), relation.getRemoteTableName(), Optional.empty()));
```
Agree, thanks for the explanation

chenjian2664 commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgresTaskFailureRecoveryTest.java:
```diff
@@ -22,4 +25,22 @@ public TestPostgresTaskFailureRecoveryTest()
+
+    @Test
+    @Override
+    protected void testDeleteWithSubquery()
```
It's surprising that the FTE at the TASK level runs successfully with the current implementation

chenjian2664 replied:
Should I copy the test from BaseFailureRecoveryTest to show here
@kokosing

chenjian2664 replied:
Feel better to show here and commented.

kokosing replied:
it could be like

```
 assertThatThrownBy(() -> super.testDeleteWithSubquery())....
 ```

chenjian2664 commented:
> Looks good. Do you think the getColumns refactor should go before or after this PR?

I am trying at I am trying at https://github.com/trinodb/trino/pull/23284. I am not sure should check the first commit to here, so how about the refactor getColumns after this PR?

kokosing commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgresTaskFailureRecoveryTest.java:
```diff
@@ -22,4 +25,38 @@ public TestPostgresTaskFailureRecoveryTest()
+    {
+        // TODO: support merge with fte
+        // The test code is identical to BaseFailureRecoveryTest#testDeleteWithSubquery
+        // The test can run successfully but the MERGE implement details is not guarantee fte, overriding to claim it
```
I don't fully understand the comment. Can you please run the MERGE query here and assert that it fails? Or is the failure not deterministic? Same below.

chenjian2664 replied:
we would get `assertThatThrownBy(super::testDeleteWithSubquery).hasMessageContaining("Expecting code to raise a throwable.");`
It's because the test defined in BaseFailureRecoveryTest can run successfully with current implementation

chenjian2664 replied:
But the `BaseJdbcFailureRecoveryTest` overwrite it and expect the test fail

chenjian2664 replied:
I’ve copied the code from `BaseFailureRecoveryTest`. Which version do you think is better? Do you prefer keeping `the assertThatThrownBy(super::testDeleteWithSubquery).hasMessageContaining("Expecting code to raise a throwable.");` ?

kokosing commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgresTaskFailureRecoveryTest.java:
```diff
@@ -22,4 +25,38 @@ public TestPostgresTaskFailureRecoveryTest()
+    @Override
+    protected void testDeleteWithSubquery()
+    {
+        // TODO: support merge with fte
```
Is there issue on github about solving this TODO? How to reproduce the problem?

chenjian2664 replied:
There is https://github.com/trinodb/trino/issues/23345 
It seems can not reproduce, the current TASK level fte for MERGE won't fail

kokosing replied:
So maybe it is now supported?

chenjian2664 replied:
Yes, seems, but it out of my understanding. I need to read what does the TASK level fte actual do. For now, from the test it says it is supported.

kokosing commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgresQueryFailureRecoveryTest.java:
```diff
@@ -22,4 +26,22 @@ public TestPostgresQueryFailureRecoveryTest()
+    protected void testDeleteWithSubquery()
+    {
+        // TODO: support merge with fte https://github.com/trinodb/trino/issues/23345
+        assertThatThrownBy(super::testDeleteWithSubquery).hasMessageContaining("expected: OptionalLong[527]\n but was: OptionalLong[0]");
```
Now I got it. It is not that is not supported, it "works" but it returns incorrect results. In other words the MERGE operation for JDBC connectors cannot be retried.

I think what we need to here is to disallow running MERGE in FTE mode for JDBC connectors now. We need to throw good error message, not incorrect results.

kokosing replied:
@losipiuk do you have any idea how connector can recognize if it is executing under FTE mode?

chenjian2664 replied:
> Now I got it. It is not that is not supported, it "works" but it returns incorrect results. In other words the MERGE operation for JDBC connectors cannot be retried.

More Interesting thing is the TASK level retry returns correct results :)

chenjian2664 replied:
There is `verifyRetryMode` in `DefaultJdbcMetadata`, maybe we could put sth in `beginMerge`
```java
if (retryMode != NO_RETRIES) {
     if (!jdbcClient.supportsMergeRetries()) {
           throw new TrinoException(NOT_SUPPORTED, "This connector does not support MERGE with retries");
     }
     if (isNonTransactionalInsert(session)) {
           throw new TrinoException(NOT_SUPPORTED, "Query and task retries are incompatible with non-transactional inserts");
     }
}
``` 
And add the method `supportsMergeRetries` in `JdbcClient`

chenjian2664 replied:
@kokosing Try above, PTAL

losipiuk replied:
> There is verifyRetryMode in DefaultJdbcMetadata, maybe we could put sth in beginMerge

Yeah - checking the `retryMode` param in `beginMerge` and acting according to that is the way to go.
I did not read the PR so I do not know if it is supposed to work out of the box with FTE (I doubt it)  or not.

kokosing commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/BasePostgresFailureRecoveryTest.java:
```diff
@@ -58,6 +59,14 @@ protected QueryRunner createQueryRunner(
+    protected void testDeleteWithSubquery()
+    {
+        assertThatThrownBy(super::testDeleteWithSubquery).hasMessageContaining("This connector does not support MERGE with retries");
+        abort("skipped");
```
no need for calling `abort`, same below

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcClient.java:
```diff
@@ -215,6 +215,8 @@ default void setTableProperties(ConnectorSession session, JdbcTableHandle handle
 
     boolean supportsRetries();
 
+    boolean supportsMergeRetries();
```
Do we need this method today? The only implementation `return false`, so let's maybe inline it for now and add it later when we know it is could be more useful.

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:
```diff
@@ -1297,6 +1297,14 @@ public ColumnHandle getMergeRowIdColumnHandle(ConnectorSession session, Connecto
+            if (!jdbcClient.supportsMergeRetries()) {
+                throw new TrinoException(NOT_SUPPORTED, "This connector does not support MERGE with retries");
+            }
+            if (isNonTransactionalInsert(session)) {
```
This condition was never checked, so it could be removed as the above `if` was always throwing an exception. It looks like we can simplify the code to:
```
if (retryMode != NO_RETRIES) {
  throw new TrinoException(NOT_SUPPORTED, "This connector does not support MERGE with fault-tolerant execution");
}
```

kokosing commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:
```diff
@@ -139,6 +142,13 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
+    @Override
+    public void testUpdateRowConcurrently()
+    {
+        abort("Postgresql doesn't support concurrent update of different columns in a row");
```
Can you do something like:
```
        assertThatThrownBy(super:: testUpdateRowConcurrently).hasMessageContaining("....");
```

chenjian2664 replied:
It fails due to the incorrect result, and the result usually is non deterministic.

chenjian2664 replied:
Tried, PTAL

kokosing replied:
> It fails due to the incorrect result

Oh... It is a big deal unfortunately. It is ok to fail with some nice error message, however no failure and incorrect results are not good.

As I understand now merge is being done on many worker nodes. Is it possible to do it on one worker only and do `commit` or `rollback` so either all UPDATE is done or none?

chenjian2664 replied:
Interesting, do u have some context about this, I mean which part I should learn for and refactor?

chenjian2664 replied:
I think you are talking is similar to the FTE, which would write the `update` and `delete` into temporary table as well

kokosing replied:
How would it work?

chenjian2664 replied:
link https://github.com/trinodb/trino/pull/19819. Here is a workaround.

chenjian2664 replied:
Will be resolved by https://github.com/trinodb/trino/pull/24075

kokosing reviewed (approved):
The last comment to go!

I like how you simplified handling FTE! Good stuff!

chenjian2664 commented:
@kokosing Thank you very much for your detailed and professional review!

chenjian2664 commented on plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -275,6 +275,26 @@ public class PostgreSqlClient
         return FULL_PUSHDOWN.apply(session, simplifiedDomain);
     };
 
+    private static final ColumnMapping POSTGRESQL_MERGE_ROW_ID_COLUMN_MAPPING = ColumnMapping.sliceMapping(
```
The `ctid` is not stable while concurrent update. we need use the primary key or unique key.

chenjian2664 replied:
Revert the impl that rely on the `ctid`, instead, using the primary keys, it would be more general but for some tables  that not defined primarykeys might not able to perform merge.

kokosing commented:
Let me know when you rebase and when it is ready for another review (merge).

kokosing commented:
CI is red

chenjian2664 commented:
> Let me know when you rebase and when it is ready for another review (merge).

PTAL, there are lots of changes in the second commits, due to the `ctid` approach is not stable when the same row is being concurrent update, change to use the primary keys.

kokosing commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:
```diff
@@ -1314,18 +1323,7 @@ public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorT
-        }
-
-        return new JdbcMergeTableHandle(handle, jdbcOutputTableHandle, primaryKeys, updateColumnChannelsBuilder.buildOrThrow());
+        return new JdbcMergeTableHandle(handle, jdbcOutputTableHandle, primaryKeys, columns, updateColumnHandles);
```
I think these changes should belong to previous commit.

kokosing commented on plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -1188,6 +1197,32 @@ protected void verifyColumnName(DatabaseMetaData databaseMetadata, String column
+            }
+            return columns.stream()
+                    .filter(column -> primaryKeyNames.contains(column.getColumnName()))
+                    .collect(toImmutableList());
```
Why do you do this filtering? To filter `ctid`?

chenjian2664 replied:
To get the primary keys from all the columns set. i,e, all column: c1, c2, c3. primarykeyNames: {c1, c2}, result: [c1, c2]

kokosing commented on plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -1188,6 +1197,32 @@ protected void verifyColumnName(DatabaseMetaData databaseMetadata, String column
     }
 
+    @Override
+    public List<JdbcColumnHandle> getPrimaryKeys(ConnectorSession session, RemoteTableName remoteTableName)
```
Why we cannot use `ctid`?

chenjian2664 replied:
The `ctid` value might be changed after the update.
see the blow logs, that I log before and after the update in the `testUpdateRowConcurrently`.
```
2024-11-28T20:36:22.057-0600	INFO	pool-104-thread-4	stdout	ctid :(0,1)
2024-11-28T20:36:22.057-0600	INFO	pool-104-thread-1	stdout	ctid :(0,1)
2024-11-28T20:36:22.057-0600	INFO	pool-104-thread-3	stdout	ctid :(0,1)
2024-11-28T20:36:22.057-0600	INFO	pool-104-thread-2	stdout	ctid :(0,1)
2024-11-28T20:36:22.380-0600	INFO	pool-104-thread-4	stdout	ctid :(0,5)
2024-11-28T20:36:22.381-0600	INFO	pool-104-thread-1	stdout	ctid :(0,5)
2024-11-28T20:36:22.383-0600	INFO	pool-104-thread-2	stdout	ctid :(0,5)
2024-11-28T20:36:22.383-0600	INFO	pool-104-thread-3	stdout	ctid :(0,5)
```

kokosing reviewed (commented):
A lot of things has changed. I need take a deeper look.

Can you please explain why updating only a subset of columns fixed the issue that MERGE is not transactional?

chenjian2664 commented:
> Can you please explain why updating only a subset of columns fixed the issue that MERGE is not transactional?

About the `update the row`, the semantic is guaranteed by the `PostgreSql`, the update lock the row implicitly.
About the `whole merge process`, I think now it's not transactional, this is as the follow up. My idea is learning the current `transactional insert`, which writing the temporary tables then perform update/delete/insert.

kokosing commented:
>  I think now it's not transactional, this is as the follow up

Should we then have MERGE be disabled behind a feature toggle till then? I am afraid that any cancellation of the MERGE may leave the table in not-coherent state. Hence I think users should turn this on on their own risk for now.

chenjian2664 commented:
> > I think now it's not transactional, this is as the follow up
> 
> Should we then have MERGE be disabled behind a feature toggle till then? I am afraid that any cancellation of the MERGE may leave the table in not-coherent state. Hence I think users should turn this on on their own risk for now.

Yes, I think it's good to have a control have it. Do you think could add session property like the `non_transactional_insert` ? Actually, if the user use the FTE mode or enable the `non_transactional_insert` we could reject it to let user know now only support non-transaction-merge

chenjian2664 commented:
> I think session property is nice here.

Add the `non-transactional-merge` session property.

kokosing commented:
`non-transactional-merge-enanbled`? Like it is on other session properties? Also see `io.trino.plugin.bigquery.BigQuerySessionProperties` there `_` are used.

kokosing reviewed (approved):
% comments

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcClient.java:
```diff
@@ -166,6 +166,8 @@ Optional<PreparedQuery> legacyImplementJoin(
 
     boolean isLimitGuaranteed(ConnectorSession session);
 
+    boolean supportsMerge();
```
PostgreSQL allows the creation of tables without primary keys. While the connector supports the MERGE operation, it cannot be performed on tables without primary keys. We need to identify such cases and provide the user with a more informative error message.That is :

1. The connector not able to MERGE, return the `This connector does not support modifying table rows`
2. The connector support MERGE, but target table without primary keys, return `The connector can not perform merge on the target table without primary keys`

kokosing commented:
Thank you! Merged.

mosabua commented:
Woooohhoooo.. congratulations @chenjian2664 -- excellent work. Also thank you so much for your help and guidance @kokosing

ebyhr commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:
```diff
@@ -128,7 +138,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
                     SUPPORTS_JOIN_PUSHDOWN_WITH_VARCHAR_EQUALITY,
                     SUPPORTS_TOPN_PUSHDOWN,
-                    SUPPORTS_TOPN_PUSHDOWN_WITH_VARCHAR -> true;
+                    SUPPORTS_TOPN_PUSHDOWN_WITH_VARCHAR,
+                    SUPPORTS_ROW_LEVEL_UPDATE,
+                    SUPPORTS_MERGE -> true;
```
Reorder as alphabetically.

ebyhr commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlJdbcConnectionAccesses.java:
```diff
@@ -97,19 +111,34 @@ public void testJdbcConnectionCreations()
+        assertJdbcConnections("MERGE INTO copy_of_customer c USING customer r ON r.custkey = c.custkey WHEN MATCHED THEN DELETE", 28, Optional.empty());
+    }
+
+    private void addPrimaryKeyToCopyTable()
```
Inline this method.

ebyhr commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlJdbcConnectionCreation.java:
```diff
@@ -100,12 +114,28 @@ public void testJdbcConnectionCreations()
+        assertJdbcConnections("MERGE INTO copy_of_customer c USING customer r ON r.custkey = c.custkey WHEN MATCHED THEN DELETE", 18, Optional.empty());
+    }
+
+    private void addPrimaryKeyToCopyTable()
```
Inline this method.

chenjian2664 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcWriteSessionProperties.java:
```diff
@@ -49,9 +50,14 @@ public JdbcWriteSessionProperties(JdbcWriteConfig writeConfig)
                         false))
                 .add(booleanProperty(
                         NON_TRANSACTIONAL_INSERT,
-                        "Do not use temporary table on insert to table",
```
It was modified by mistake, actually should update the description for `NON_TRANSACTIONAL_MERGE`

chenjian2664 replied:
sent https://github.com/trinodb/trino/pull/24431

shohamyamin commented:
@chenjian2664 this PR covers support MERGE for oracle as well?

chenjian2664 commented:
@shohamyamin This pr not covers oracle, but I am working on it

chenjian2664 commented:
The oracle work is tracking in https://github.com/trinodb/trino/pull/23034 @shohamyamin FYI

shohamyamin commented:
@chenjian2664 where is the pr for the oracle merge support(you write by mistake this pr)

ssheikin commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:
```diff
@@ -138,6 +139,7 @@ public CachingJdbcClient(
         columnsCache = buildCache(ticker, cacheMaximumSize, metadataCachingTtl);
         tableCommentsCache = buildCache(ticker, cacheMaximumSize, metadataCachingTtl);
         statisticsCache = buildCache(ticker, cacheMaximumSize, statisticsCachingTtl);
+        tablePrimaryKeysCache = buildCache(ticker, cacheMaximumSize, statisticsCachingTtl);
```
@chenjian2664 I wonder if this cache should use `metadataCachingTtl` instead of `statisticsCachingTtl`
